module_3/gifts/controller/src/main/java/com/epam/esm/controller/CertificateController.java

```diff
@@ -0,0 +1,148 @@
+		addLinksToCertificate(createdCertificateDto);
+		return createdCertificateDto;
+	}
+
+	/**
+	 * Reads certificate with passed id.
+	 * 
+	 * @param id the id of certificate to be read
+	 * @return certificate with passed id
+	 */
+	@GetMapping("/{id}")
+	@ResponseStatus(HttpStatus.OK)
+	public CertificateDto readById(@PathVariable long id) {
+		CertificateDto certificateDto = certificateService.readById(id);
+		addLinksToCertificate(certificateDto);
+		return certificateDto;
+	}
+
+	/**
+	 * Reads all certificates according to passed parameters.
+	 * 
+	 * @param params the parameters which define the choice of certificates and
+	 *               their ordering
+	 * @return certificates which meet passed parameters
+	 */
+	@GetMapping
+	public ResponseEntity<List<CertificateDto>> readAll(@RequestParam MultiValueMap<String, String> params) {
+		List<CertificateDto> certificates = certificateService.readAll(params);
+		if (certificates == null || certificates.isEmpty()) {
+			return new ResponseEntity<>(certificates, HttpStatus.NO_CONTENT);
+		} else {
+			for (CertificateDto certificateDto : certificates) {
+				addLinksToCertificate(certificateDto);
+			}
+			return new ResponseEntity<>(certificates, HttpStatus.OK);
+		}
+	}
+
+	/**
+	 * Updates certificate fields with passed id using not {@code null} fields of
+	 * passed certificate entity.
+	 * 
+	 * @param id          the id of certificate to be updated
+	 * @param certificate certificate entity which contains fields with new values
+	 *                    to be set
+	 * @return updated certificate
+	 */
+	@PatchMapping("/{id}")
+	@ResponseStatus(HttpStatus.OK)
+	public CertificateDto updateCertificateFields(@RequestBody CertificateDto certificate, @PathVariable long id) {
+		CertificateDto updatedCertificateDto = certificateService.updateCertificateFields(id, certificate);
+		addLinksToCertificate(updatedCertificateDto);
+		return updatedCertificateDto;
+	}
+
+	/**
+	 * Updates entire certificate with passed id using all fields of passed
+	 * certificate.
+	 * 
+	 * @param id          the id of certificate to be updated
+	 * @param certificate certificate entity which contains fields with new values
+	 *                    to be set
+	 * @return updated certificate
+	 */
+	@PutMapping("/{id}")
+	@ResponseStatus(HttpStatus.OK)
+	public CertificateDto updateEntireCertificate(@RequestBody CertificateDto certificate, @PathVariable long id) {
+		CertificateDto updatedCertificateDto = certificateService.updateEntireCertificate(id, certificate);
+		addLinksToCertificate(updatedCertificateDto);
+		return updatedCertificateDto;
+	}
+
+	/**
+	 * Deletes certificate with passed id.
+	 * 
+	 * @param id the id of certificate to be deleted
+	 *
+	 */
+	@DeleteMapping("/{id}")
+	@ResponseStatus(HttpStatus.NO_CONTENT)
+	public void delete(@PathVariable long id) {
+		certificateService.delete(id);
+	}
+
+	public static void addLinksToCertificate(CertificateDto certificateDto) {
```

Comment:
@LizavetaSinitsyna I think this method could be `privet`.

Reply:
@AliaksandrLiulis, the reason why I left this method public is that it is used in OrderController as Order entity contains certificates. So it need to be public in order to not repeat the same part of code.